packages/order-watcher/src/order_watcher/order_watcher_websocket.ts

```diff
@@ -0,0 +1,173 @@
+            connection.on('message', await this._onMessageCallbackAsync.bind(this, connection));
+            connection.on('close', this._onCloseCallback.bind(this, connection));
+            this._connectionStore.add(connection);
+        });
+
+        // Have the WebSocket server subscribe to the OrderWatcher to receive updates.
+        // These updates are then broadcast to clients in the _connectionStore.
+        const broadcastCallback: OnOrderStateChangeCallback = this._broadcastCallback.bind(this);
+        this._orderWatcher.subscribe(broadcastCallback);
+    }
+
+    /**
+     * Activates the WebSocket server by having its HTTP server start listening.
+     */
+    public listen(): void {
+        const port = process.env.ORDER_WATCHER_HTTP_PORT || DEFAULT_HTTP_PORT;
+        this._httpServer.listen(port, () => {
+            logUtils.log(`${new Date()} [Server] Listening on port ${port}`);
+        });
+    }
+
+    /**
+     * Deactivates the WebSocket server by stopping the HTTP server from accepting
+     * new connections.
+     */
+    public close(): void {
+        this._httpServer.close();
+    }
+
+    private async _onMessageCallbackAsync(connection: WebSocket.connection, message: any): Promise<void> {
+        const response: WebSocketResponse = {
+            action: null,
+            success: false,
+            result: null,
+        };
+        try {
+            assert.doesConformToSchema('message', message, webSocketUtf8MessageSchema);
+            const request: WebSocketRequest = JSON.parse(message.utf8Data);
+            assert.doesConformToSchema('request', request, webSocketRequestSchema);
+            response.action = request.action;
+            response.success = true;
+            response.result = await this._routeRequestAsync(request);
+        } catch (err) {
+            response.result = err.toString();
+        }
+        logUtils.log(`${new Date()} [Server] OrderWatcher output: ${JSON.stringify(response)}`);
+        connection.sendUTF(JSON.stringify(response));
+    }
+
+    private _onCloseCallback(connection: WebSocket.connection): void {
+        this._connectionStore.delete(connection);
+        logUtils.log(`${new Date()} [Server] Client ${connection.remoteAddress} disconnected.`);
+    }
+
+    private async _routeRequestAsync(request: WebSocketRequest): Promise<any> {
+        logUtils.log(`${new Date()} [Server] Request received: ${request.action}`);
+        let result = null;
+        switch (request.action) {
+            case OrderWatcherAction.AddOrder: {
+                const signedOrder: SignedOrder = OrderWatcherWebSocketServer._parseSignedOrder(request.signedOrder);
+                await this._orderWatcher.addOrderAsync(signedOrder);
+                break;
+            }
+            case OrderWatcherAction.RemoveOrder: {
+                const orderHash = request.orderHash || '_';
```

Comment:
This looks like a hack. Can you explain?

Reply:
Yeah. I'm trying to get around the type checker. Because request.orderHash could be null if it's an ADD_ORDER request, the type of request.orderHash is string or null, and this can't be passed to orderHash which expects a string. Ideally the type checker would know we're guarded by OrderWatcherAction.RemoveOrder and so request.orderHash will definitely not be null. Open to better workarounds too. This is a common problem I face with typescript.

Reply:
Yeah, if you do the union type suggestion I also commented on, this issue will go away.